src/serializer/ResidualHeapSerializer.js

```diff
@@ -718,23 +720,50 @@ export class ResidualHeapSerializer {
     return Array.from(result);
   }
 
-  // Determine if a value is effectively referenced by a single additional function.
+  isDefinedInsideFunction(childFunction: FunctionValue, maybeParentFunctions: Set<FunctionValue>): boolean {
+    for (let maybeParentFunction of maybeParentFunctions) {
+      if (childFunction === maybeParentFunction) {
+        continue;
+      }
+      let env = childFunction.$Environment;
+      while (env.parent !== null) {
+        let envRecord = env.environmentRecord;
+        if (envRecord instanceof FunctionEnvironmentRecord && envRecord.$FunctionObject === maybeParentFunction) {
+          return true;
+        }
+        env = env.parent;
+      }
+    }
+    return false;
+  }
+
+  // Determine if a value is effectively referenced by an optimized function.
```

Comment:
If my interpretation is right, then this function now returns the single root optimized function (even if there are other child optimized functions). So maybe rename function to tryGetOptimizedFunctionRoot.